logstash-core/src/main/java/org/logstash/LogstashJavaCompat.java

```diff
@@ -0,0 +1,72 @@
+package org.logstash;
+
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.Collection;
+import org.codehaus.janino.ClassBodyEvaluator;
+import org.logstash.ackedqueue.io.ByteBufferCleaner;
+
+/**
+ * Logic around ensuring compatibility with Java 8 and 9 simultaneously.
+ */
+public final class LogstashJavaCompat {
+
+    /**
+     * True if current JVM is a Java 9 implementation.
+     */
+    public static final boolean IS_JAVA_9_OR_GREATER = isAtLeastJava9();
+
+    /**
+     * Sets up an appropriate implementation of {@link ByteBufferCleaner} depending no whether or
+     * not the current JVM is a Java 9 implementation.
+     * @return ByteBufferCleaner
+     */
+    public static ByteBufferCleaner setupBytebufferCleaner() {
+        final ClassBodyEvaluator se = new ClassBodyEvaluator();
+        final Collection<String> imports = new ArrayList<>();
+        imports.add("java.nio.MappedByteBuffer");
+        final String cleanerCode;
+        final String ctorCode;
+        final String fieldsCode;
+        if (isAtLeastJava9()) {
+            imports.add("sun.misc.Unsafe");
+            imports.add("java.lang.reflect.Field");
+            cleanerCode = "unsafe.invokeCleaner(buffer);";
+            ctorCode = "Field unsafeField = Unsafe.class.getDeclaredField(\"theUnsafe\");" +
+                "unsafeField.setAccessible(true);" +
```

Comment:
I don't think reflection hacks are needed here. I _think_ the cleaner API has been replaced in Java 9 by the [Cleaner](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html) object.

Reply:
@jakelandis Not sure what the `Cleaner` gets me here, I will still need to implement an appropriate `Runnable` for it to trigger and that will execute this exact action won't I?

Reply:
I see now `Cleaner` is simply a newer/better version of finalize, still coming up to speed on Java9. 
Man is the history behind this quite long, led mostly by lucene https://issues.apache.org/jira/browse/LUCENE-6989 and long discussions on the mailing list http://mail.openjdk.java.net/pipermail/core-libs-dev/2016-January/thread.html#38243. 

Also a long discussion for Java 10 : https://bugs.openjdk.java.net/browse/JDK-4724038 

anyway... this change is identical in function (but not form) to Lucene so +1

Sorry for the noise.